Swarm: conditional hand-off transition improvements

marklysze opened the pull request:
## Why are these changes needed?

Feature #1:
For Swarm ON_CONDITION hand-offs, they are always included in the agent's list of available functions. To provide more control over whether they are available, an `available` parameter is added. This can be set to a function that returns a boolean, or the value of a boolean context value. If the available parameter is not provided, it will always be available.

Feature #2:
Provides the ability to define different AFTER_WORK options based on a function that returns an AfterWorkOption or a SwarmAgent. Previously the AFTER_WORK could take a Callable but it was restricted to returning a SwarmAgent.

NOTE: This PR **requires PR #104 to be merged** as that has changes to context variable handling and agent state update hooks, which this utilises.

## Related issue number

#30 - Feature number 1: Enable/Disable hand-offs conditionally for ON_CONDITION
#30 - Feature number 2: Function-based AFTER_WORKS

## Checks

- [ ] I've included any doc changes needed for https://ag2ai.github.io/ag2/. See https://ag2ai.github.io/ag2/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

marklysze commented:
Example program:
```
from typing import Any, Dict, List
from autogen import initiate_swarm_chat, SwarmAgent, SwarmResult, ON_CONDITION, AFTER_WORK, AfterWorkOption
import os

llm_config = {
    "config_list":
    [
        {
            "api_type": "openai",
            "model": "gpt-4o-mini",
            "api_key": os.getenv("OPENAI_API_KEY"),
            "cache_seed": None
        }
    ]
}


context_variables = {
    "agent_1_finished": False, 
    "agent_2_finished": False, 
    "agent_3_finished": False,
}

def do_work_for_a(context_variables: Dict[str, Any]) -> SwarmResult:
    """Run this to do your work"""
    context_variables["agent_1_finished"] = True
    return SwarmResult(context_variables=context_variables, values="Work has been done for a")

def do_work_for_b(context_variables: Dict[str, Any]) -> SwarmResult:
    """Run this to do your work"""
    context_variables["agent_2_finished"] = True
    return SwarmResult(context_variables=context_variables, values="Work has been done for b")

def do_work_for_c(context_variables: Dict[str, Any]) -> SwarmResult:
    """Run this to do your work"""
    context_variables["agent_3_finished"] = True
    return SwarmResult(context_variables=context_variables, values="Work has been done for c")

agent_1 = SwarmAgent(
    name="Agent_No_1",
    system_message="You are a helpful assistant. Ensure you do your work a.",
    llm_config=llm_config,
    functions=[do_work_for_a]
)

agent_2 = SwarmAgent(
    name="Agent_No_2",
    system_message="You are a helpful assistant. Ensure you do your work b.",
    llm_config=llm_config,
    functions=[do_work_for_b]
)

agent_3 = SwarmAgent(
    name="Agent_No_3",
    system_message="You are a helpful assistant. Ensure you do your work c.",
    llm_config=llm_config,
    functions=[do_work_for_c]
)

agent_4 = SwarmAgent(
    name="Agent_No_4",
    system_message="You are a helpful assistant.",
    llm_config=llm_config,
)

def transfer_if_work_done_for_b(agent: SwarmAgent, messages: List[Dict[str, Any]]) -> bool:
    return agent.get_context("agent_2_finished")

agent_1.register_hand_off(
    hand_to=[
        ON_CONDITION(target=agent_2, condition="After you've done your work, go to agent 2", available="agent_1_finished"),
        ON_CONDITION(target=agent_3, condition="You should never need this, but if you see it, run it!", available="agent_2_finished"),
        ON_CONDITION(target=agent_3, condition="You should never need this, but if you see it, run it!", available=transfer_if_work_done_for_b),
        ON_CONDITION(target=agent_4, condition="You should never need this, but if you see it, run it!", available="agent_3_finished"),
        ON_CONDITION(target=agent_4, condition="Never call this function although it's always available"),
        AFTER_WORK(agent=agent_4)
    ]
)

agent_2.register_hand_off(
    hand_to=[
        ON_CONDITION(target=agent_3, condition="After you've done your work, go to agent 3", available="agent_2_finished"),
        ON_CONDITION(target=agent_4, condition="You should never need this, but if you see it, run it!", available="agent_3_finished"),
        AFTER_WORK(agent=agent_4)
    ]
)

agent_3.register_hand_off(
    hand_to=[
        ON_CONDITION(target=agent_4, condition="Go to the next agent, agent 4", available="agent_3_finished"),
        AFTER_WORK(AfterWorkOption.TERMINATE)
    ]
)

# Initiate the swarm chat
chat_result, context_vars, last_speaker = initiate_swarm_chat(
    initial_agent=agent_1,
    messages="Do your work agents!",
    agents=[agent_1, agent_2, agent_3, agent_4],
    context_variables=context_variables,
)

# Print the results
print("\n---- Chat History ----")
for message in chat_result.chat_history:
    if "name" in message:
        print(f"{message['name']}: {message.get('content', '')}")
    else:
        print(f"{message.get('content', '')}")

print("\n---- Final Context Variables ----")
for key, value in context_vars.items():
    print(f"{key}: {value}")
```

marklysze commented:
Feature #2 example program:
```
from typing import Any, Dict, List, Union
from autogen import initiate_swarm_chat, SwarmAgent, SwarmResult, ON_CONDITION, AFTER_WORK, AfterWorkOption, GroupChat
import os

llm_config = {
    "config_list":
    [
        {
            "api_type": "openai",
            "model": "gpt-4o-mini",
            "api_key": os.getenv("OPENAI_API_KEY"),
            "cache_seed": None
        }
    ]
}


context_variables = {
    "agent_1_finished": False, 
    "agent_2_finished": False, 
    "agent_3_finished": False,
}

def do_work_for_a(context_variables: Dict[str, Any]) -> SwarmResult:
    """Run this to do your work"""
    context_variables["agent_1_finished"] = True
    return SwarmResult(context_variables=context_variables, values="Work has been done for a")

def do_work_for_b(context_variables: Dict[str, Any]) -> SwarmResult:
    """Run this to do your work"""
    context_variables["agent_2_finished"] = True
    return SwarmResult(context_variables=context_variables, values="Work has been done for b")

def do_work_for_c(context_variables: Dict[str, Any]) -> SwarmResult:
    """Run this to do your work"""
    context_variables["agent_3_finished"] = True
    return SwarmResult(context_variables=context_variables, values="Work has been done for c")

agent_1 = SwarmAgent(
    name="Agent_No_1",
    system_message="You are a helpful assistant. Ensure you do your work a.",
    llm_config=llm_config,
    functions=[do_work_for_a]
)

agent_2 = SwarmAgent(
    name="Agent_No_2",
    system_message="You are a helpful assistant. Ensure you do your work b.",
    llm_config=llm_config,
    functions=[do_work_for_b]
)

agent_3 = SwarmAgent(
    name="Agent_No_3",
    system_message="You are a helpful assistant. Ensure you do your work c.",
    llm_config=llm_config,
    functions=[do_work_for_c]
)

agent_4 = SwarmAgent(
    name="Agent_No_4",
    system_message="You are a helpful assistant, but will never get called.",
    llm_config=llm_config,
)

def transfer_if_work_done_for_b(agent: SwarmAgent, messages: List[Dict[str, Any]]) -> bool:
    return agent.get_context("agent_2_finished")

def evaluate_next_step(last_speaker: SwarmAgent, messages: List[Dict[str, Any]], groupchat: GroupChat) -> Union[AfterWorkOption, SwarmAgent, str]:
        if not last_speaker.get_context("agent_1_finished"):
             return agent_1
        if not last_speaker.get_context("agent_2_finished"):
             return agent_2
        if not last_speaker.get_context("agent_3_finished"):
             return "Agent_No_3"
        return AfterWorkOption.TERMINATE

agent_1.register_hand_off(
    hand_to=[
        AFTER_WORK(evaluate_next_step)
    ]
)

agent_2.register_hand_off(
    hand_to=[
        AFTER_WORK(evaluate_next_step)
    ]
)

def evaluate_last_after_work(last_speaker: SwarmAgent, messages: List[Dict[str, Any]], groupchat: GroupChat) -> Union[AfterWorkOption, SwarmAgent, str]:
    if not last_speaker.get_context("agent_1_finished"):
            return agent_1
    if not last_speaker.get_context("agent_2_finished"):
            return agent_2
    if not last_speaker.get_context("agent_3_finished"):
            return "Agent_No_3"
    return AfterWorkOption.TERMINATE

# Initiate the swarm chat
chat_result, context_vars, last_speaker = initiate_swarm_chat(
    initial_agent=agent_1,
    messages="Do your work agents!",
    agents=[agent_1, agent_2, agent_3, agent_4],
    context_variables=context_variables,
    after_work=evaluate_last_after_work
)

# Print the results
print("\n---- Chat History ----")
for message in chat_result.chat_history:
    if "name" in message:
        print(f"{message['name']}: {message.get('content', '')}")
    else:
        print(f"{message.get('content', '')}")

print("\n---- Final Context Variables ----")
for key, value in context_vars.items():
    print(f"{key}: {value}")
```
Output:
```
_User (to chat_manager):

Do your work agents!

--------------------------------------------------------------------------------

Next speaker: Agent_No_1

Agent_No_1 (to chat_manager):

***** Suggested tool call (call_cynxXyY6d2rAu8DDaDZS5VX4): do_work_for_a *****
Arguments: 
{}
******************************************************************************

--------------------------------------------------------------------------------

Next speaker: Tool_Execution


>>>>>>>> EXECUTING FUNCTION do_work_for_a...
Tool_Execution (to chat_manager):

***** Response from calling tool (call_cynxXyY6d2rAu8DDaDZS5VX4) *****
Work has been done for a
**********************************************************************

--------------------------------------------------------------------------------

Next speaker: Agent_No_1

Agent_No_1 (to chat_manager):

The work has been completed! If you need anything else, feel free to ask.

--------------------------------------------------------------------------------

Next speaker: Agent_No_2

Agent_No_2 (to chat_manager):

***** Suggested tool call (call_KVlGmiTvGW4Pfft7VAbIjU3C): do_work_for_b *****
Arguments: 
{}
******************************************************************************

--------------------------------------------------------------------------------

Next speaker: Tool_Execution


>>>>>>>> EXECUTING FUNCTION do_work_for_b...
Tool_Execution (to chat_manager):

***** Response from calling tool (call_KVlGmiTvGW4Pfft7VAbIjU3C) *****
Work has been done for b
**********************************************************************

--------------------------------------------------------------------------------

Next speaker: Agent_No_2

Agent_No_2 (to chat_manager):

The work has been completed for both tasks! If you need anything else, feel free to ask.

--------------------------------------------------------------------------------

Next speaker: Agent_No_3

Agent_No_3 (to chat_manager):

***** Suggested tool call (call_YhtMy4Ei6xr6BErgjaxIuQvj): do_work_for_c *****
Arguments: 
{}
******************************************************************************

--------------------------------------------------------------------------------

Next speaker: Tool_Execution


>>>>>>>> EXECUTING FUNCTION do_work_for_c...
Tool_Execution (to chat_manager):

***** Response from calling tool (call_YhtMy4Ei6xr6BErgjaxIuQvj) *****
Work has been done for c
**********************************************************************

--------------------------------------------------------------------------------

Next speaker: Agent_No_3

Agent_No_3 (to chat_manager):

All tasks have been completed successfully! If you need any further assistance, feel free to ask.

--------------------------------------------------------------------------------

---- Chat History ----
Do your work agents!
Agent_No_1: None
Tool_Execution: Work has been done for a
Agent_No_1: The work has been completed! If you need anything else, feel free to ask.
Agent_No_2: None
Tool_Execution: Work has been done for b
Agent_No_2: The work has been completed for both tasks! If you need anything else, feel free to ask.
Agent_No_3: None
Tool_Execution: Work has been done for c
Agent_No_3: All tasks have been completed successfully! If you need any further assistance, feel free to ask.

---- Final Context Variables ----
agent_1_finished: True
agent_2_finished: True
agent_3_finished: True
```

marklysze commented:
@yiranwu0 if you're able to help with implementing # 4 feature from #193 that would be great.

yiranwu0 commented on autogen/agentchat/contrib/swarm_agent.py:
```diff
@@ -387,6 +405,17 @@ def __init__(
+        if name != __TOOL_EXECUTOR_NAME__:
+            self.register_hook("update_agent_state", self._update_agent_state_hook)
+
+    def _update_agent_state_hook(self, agent: Agent, messages: Optional[List[Dict]] = None) -> None:
+        """Hook to update the agent state and update conditional functions."""
+        self._update_conditional_functions()
```
Why calling another function in a function? How about update system messages?
I think we can directly put _update_conditional_functions and update system messages here if they are not used any else where.

yiranwu0 commented on autogen/agentchat/contrib/swarm_agent.py:
```diff
@@ -191,28 +197,36 @@ def determine_next_agent(last_speaker: SwarmAgent, groupchat: GroupChat):
-                return last_speaker
-        elif isinstance(tmp_after_work, Callable):
-            return tmp_after_work(last_speaker, groupchat.messages, groupchat, context_variables)
+        # Resolve after_work condition (agent-level overrides global)
+        after_work_condition = (
+            last_swarm_speaker.after_work if last_swarm_speaker.after_work is not None else after_work
+        )
+        if isinstance(after_work_condition, AFTER_WORK):
+            after_work_condition = after_work_condition.agent
+
+        # Evaluate callable after_work
+        if isinstance(after_work_condition, Callable):
+            after_work_condition = after_work_condition(last_speaker, groupchat.messages, groupchat)
+
+        if isinstance(after_work_condition, str):  # Agent name in a string
+            if after_work_condition in swarm_agent_names:
+                after_work_condition = groupchat.agent_by_name(name=after_work_condition)
+            else:
+                raise ValueError(f"Invalid agent name in after_work: {after_work_condition}")
+
+        # Determine next action based on after_work_condition
+        if isinstance(after_work_condition, SwarmAgent):
+            return after_work_condition
+        elif after_work_condition == AfterWorkOption.TERMINATE or (
+            user_agent is None and after_work_condition == AfterWorkOption.REVERT_TO_USER
+        ):
+            return None
+        elif after_work_condition == AfterWorkOption.REVERT_TO_USER:
+            return user_agent
+        elif after_work_condition == AfterWorkOption.STAY:
+            return last_speaker
```
What does this change do? I just see the code is growing longer with the elif changed to if, and the name change. The original code is much cleaner and I think we should modify from there.

marklysze commented:
Thanks for the run through @yiranwu0, I've tidied the code as per discussion. If you get a chance to look over that would be great :)

marklysze commented:
@yiranwu0, I fixed a couple of bugs in the code (found while creating notebook):
- conditional hands-off weren't included for nested chats as they should have been
- "message" in nested chats could be overridden and not restored, fixed
- group chat manager will also have the context variables so they can be accessed from functions within the nested chat (as that's the agent that is the sender)

Can you please look over and let me know if we can merge.

BabyCNM commented on autogen/agentchat/contrib/swarm_agent.py:
```diff
@@ -47,6 +49,7 @@ def __post_init__(self):
 class ON_CONDITION:
     target: Union["SwarmAgent", Dict[str, Any]] = None
     condition: str = ""
+    available: Optional[Union[Callable, str]] = None
```
Should we add a brief documentation to introduce both "AFTER_WORK" and "ON_CONDITION"?

BabyCNM replied:
A few lines of docstring might be sufficient already.

BabyCNM commented on autogen/agentchat/contrib/swarm_agent.py:
```diff
@@ -191,34 +192,41 @@ def determine_next_agent(last_speaker: SwarmAgent, groupchat: GroupChat):
+                raise ValueError(f"Invalid agent name in after_work: {after_work_condition}")
+        elif isinstance(after_work_condition, SwarmAgent):
+            return after_work_condition
+        elif after_work_condition == AfterWorkOption.TERMINATE or (
```
Maybe a nested if statement with "AfterWorkOption" and then iterate through the options might be more readable?

BabyCNM commented on autogen/agentchat/contrib/swarm_agent.py:
```diff
@@ -191,34 +192,41 @@ def determine_next_agent(last_speaker: SwarmAgent, groupchat: GroupChat):
+            after_work_condition = after_work_condition.agent
+
+        # Evaluate callable after_work
+        if isinstance(after_work_condition, Callable):
```
Maybe "elif" is more readable here?

BabyCNM commented:
I am not familiar with the use cases of Swarm Agent, but this PR looks good to me. I have a few lint comments, and feel free to ignore completely or address.

marklysze commented:
Thanks again @BabyCNM and @yiranwu0, I'll merge these changes once all tests have passed.
